### PR TITLE
Fix bug in Tar preventing extraction of hardlinks or entries starting with `.\`

### LINKED ISSUE #68360: Extract_HardLinkEntry_TargetInsideDirectory test from System.Formats.Tar.Tests fails on Android/iOS/tvOS
Newly added test from https://github.com/dotnet/runtime/pull/67883 fails on runtime-extra-platfroms.
I haven't investigated whether it is expected, just filing this so I can disable the test with ActiveIssue.

Android fails with 
```
System.UnauthorizedAccessException : Access to the path '/data/user/0/net.dot.System.Formats.Tar.Tests/cache/qtsmyi3v.nmj/link' is denied.
---- System.IO.IOException : Permission denied


Stack trace
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Interop.CheckIo(Int64 result, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Interop.CheckIo(Int32 result, String path, Boolean isDirectory, Func`2 errorRewriter)
   at System.Formats.Tar.TarEntry.ExtractAsHardLink(String targetFilePath, String hardLinkFilePath)
   at System.Formats.Tar.TarEntry.ExtractToFileInternal(String filePath, String linkTargetPath, Boolean overwrite)
   at System.Formats.Tar.TarEntry.ExtractRelativeToDirectory(String destinationDirectoryPath, Boolean overwrite)
   at System.Formats.Tar.TarFile.ExtractToDirectoryInternal(Stream source, String destinationDirectoryPath, Boolean overwriteFiles, Boolean leaveOpen)
   at System.Formats.Tar.TarFile.ExtractToDirectory(Stream source, String destinationDirectoryName, Boolean overwriteFiles)
   at System.Formats.Tar.Tests.TarFile_ExtractToDirectory_Stream_Tests.Extract_LinkEntry_TargetInsideDirectory_Internal(TarEntryType entryType) in /_/src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs:line 123
   at System.Formats.Tar.Tests.TarFile_ExtractToDirectory_Stream_Tests.Extract_HardLinkEntry_TargetInsideDirectory() in /_/src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs:line 101
   at System.Reflection.RuntimeMethodInfo.InvokeNonEmitUnsafe(Object obj, IntPtr* byrefParameters, Span`1 argsForTemporaryMonoSupport, BindingFlags invokeAttr)
----- Inner Stack Trace -----
```

iOS/tvOS Simulators fail with

```
System.IO.IOException : Extracting the Tar entry '/Users/helix-runner/Library/Developer/CoreSimulator/Devices/11416593-2690-443B-964F-E9F6694BA19A/dat' would have resulted in a link target outside the specified destination directory: '/Users/helix-runner/Library/Developer/CoreSimulator/Devices/11416593-2690-443B-964F-E9F6694BA19A/data/Containers/Data/Application/1BD07C3B-0285-4EAB-943A-2BBFDD9E63A5/tmp/f2gsvib5.4m4/'


Stack trace
   at System.Formats.Tar.TarEntry.<ExtractRelativeToDirectory>g__GetSanitizedFullPath|34_0(String destinationDirectoryFullPath, String path, String exceptionMessage)
   at System.Formats.Tar.TarEntry.ExtractRelativeToDirectory(String destinationDirectoryPath, Boolean overwrite)
   at System.Formats.Tar.TarFile.ExtractToDirectoryInternal(Stream source, String destinationDirectoryPath, Boolean overwriteFiles, Boolean leaveOpen)
   at System.Formats.Tar.TarFile.ExtractToDirectory(Stream source, String destinationDirectoryName, Boolean overwriteFiles)
   at System.Formats.Tar.Tests.TarFile_ExtractToDirectory_Stream_Tests.Extract_LinkEntry_TargetInsideDirectory_Internal(TarEntryType entryType)
   at System.Formats.Tar.Tests.TarFile_ExtractToDirectory_Stream_Tests.Extract_HardLinkEntry_TargetInsideDirectory()
   at System.Reflection.RuntimeMethodInfo.InvokeNonEmitUnsafe(Object obj, IntPtr* byrefParameters, Span`1 argsForTemporaryMonoSupport, BindingFlags invokeAttr)
```

FYI @carlossanlop

### LINKED ISSUE #70516: Tar entries should be able to handle paths that begin with "./", including the root path itself
### Description

Bug found while investigating this issue: https://github.com/dotnet/runtime/issues/70509 

`TarReader.GetNextEntry` is unable to retrieve entries that have the path set to `./`.

### Reproduction Steps

Download https://dotnetcli.azureedge.net/dotnet/Runtime/6.0.5-servicing.22213.9/dotnet-runtime-6.0.5-osx-x64.tar.gz

```cs
string targetPath = "D:/some/destination";
using FileStream compressedStream = File.OpenRead(@"D:/path/to/dotnet-runtime-6.0.5-osx-x64.tar.gz");
using GZipStream decompressor = new(compressedStream, CompressionMode.Decompress);
TarFile.ExtractToDirectory(source: decompressor, destinationDirectoryName: targetPath, overwriteFiles: true);
```

### Expected behavior

The archive is extracted.

### Actual behavior

```
    System.IO.IOException : Extracting the Tar entry './' would have resulted in a file outside the specified destination directory: 'C:\Users\carlos\AppData\Local\Temp\zm5gtquh.qu5\'
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by carlossanlop**
Fixes https://github.com/dotnet/runtime/issues/68360 
Fixes https://github.com/dotnet/runtime/issues/70516

Depends on a new asset file: https://github.com/dotnet/runtime-assets/pull/247

Two bug fixes in one PR:
- Hardlinks were not being extracted sometimes because the target path was detected as being outside the destination directory.
- Entries that either start with `.\` or have a final segment literally named `.\`, were failing to extract because they were detected as being outside the destination directory.

There were two root causes, both happening when generating the sanitized path:
- If there was a trailing separator, it was not being handled properly, causing the code to either think the root path was not the same, or not appending it to the rest of the entry's path correctly.
- The `.\` segment is a redundant segment, no matter where it is located. It should get removed always, and we should use `GetFullPath` for that.

### CODE REVIEW

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs
Comment (eerhardt) [inline src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs:15]: How do you know you are unelevated here? Don't we run CI tests as `sudo`?

File: src/libraries/System.Formats.Tar/tests/TarTestsBase.cs
Comment (eerhardt) [inline src/libraries/System.Formats.Tar/tests/TarTestsBase.cs:117]: Can this be refactored with the code above it to reduce duplication?

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Tests.cs
Comment (eerhardt) [inline src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Tests.cs:560]: Does this test provide any value over ExtractEntriesWithSlashDotPrefix? It appears to be a subset of that test.

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs:15]: It's my understanding that we only run as sudo when we explicitly wrap the code with RemoteExecutor, and that has been the behavior I've seen in the few tests I've protected with RemoteExecutor.

Is that assumption incorrect?

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Tests.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Tests.cs:560]: I guess not. I wrote this test first, to make sure my fix was working and not throwing when an entry was _only_ read.

Then I wrote the other one, which would not just read, but also would extract the entry, and that's where the exception was originally happening.

I'll remove this test.

File: src/libraries/System.Formats.Tar/tests/TarTestsBase.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarTestsBase.cs:117]: Sure.

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs
Comment (eerhardt) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs:15]: I'm not sure. I've seen other tests be conditional on whether the process is elevated or not. For example:

https://github.com/dotnet/runtime/blob/ab0de531c89f87eca90899a47360fe2cb73e2d02/src/libraries/System.Net.Sockets/tests/FunctionalTests/CreateSocketTests.cs#L43-L44

https://github.com/dotnet/runtime/blob/ab0de531c89f87eca90899a47360fe2cb73e2d02/src/libraries/System.Net.Sockets/tests/FunctionalTests/CreateSocketTests.cs#L98-L121

I'm not sure of all the ways our tests are run. So if you are expecting this test to only work in an unelevated process, it might be good to stick that in a condition.

Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs:15]: BTW this test already existed, I merely renamed the file and the test class.

Comment (eerhardt) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs:15]: ok - if it is already passing in CI, then we can just keep it.

**Comment by carlossanlop** [comment]
@ViktorHofer do you mind scheduling a runtime-assets dependency update? The CI is failing because it cannot find the new dependency that was merged here: https://github.com/dotnet/runtime-assets/pull/247

```
System.IO.DirectoryNotFoundException : Could not find a part of the path '/private/tmp/helix/working/9A3C08AC/w/9AE80860/e/strange/prefixDotSlashAndCurrentFolderEntry.tar'.
```

(I have been testing this locally with a hardcoded nuget package I generated.)

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs:15]: Cool.

BTW the last argument of the `RemoteExecutor.Invoke` method is where I indicate that I want the code to be elevated:

```cs
RemoteExecutor.Invoke(() =>
            {
            // Code to elevate here
            }, new RemoteInvokeOptions { RunAsSudo = true }).Dispose();
```

**Comment by ViktorHofer** [comment]
> @ViktorHofer do you mind scheduling a runtime-assets dependency update?

These happen automatically as part of dependency updates. That's the PR that updates the runtime-assets artifacts in dotnet/runtime: https://github.com/dotnet/runtime/pull/70476. The PR seems to be blocked on something. Might make sense to ping for a status update or look at the failures yourself.

Alternatively you can update the package version yourself in this PR.

**Comment by eerhardt** [comment]
> Alternatively you can update the package version yourself in this PR.

that seems like the best route to unblock this PR.

**Comment by carlossanlop** [comment]
CI failures are unrelated.

**Review by eerhardt — APPROVED** [review]
LGTM. :shipit:

**Pull request merged**
